kernel/sched.c

```diff
@@ -489,3 +497,44 @@ int _impl_k_is_preempt_thread(void)
 #ifdef CONFIG_USERSPACE
 _SYSCALL_HANDLER0_SIMPLE(k_is_preempt_thread);
 #endif
+
+#ifdef CONFIG_SMP
+int _get_highest_ready_prio(void)
+{
+	int p;
+
+	for (p = 0; p < ARRAY_SIZE(_kernel.ready_q.q); p++) {
+		if (!sys_dlist_is_empty(&_kernel.ready_q.q[p])) {
+			break;
+		}
+	}
+
+	__ASSERT(p < K_NUM_PRIORITIES, "No ready prio");
+
+	return p - _NUM_COOP_PRIO;
+}
+
+struct k_thread *_get_next_ready_thread(void)
+{
+	int p, mycpu = _arch_curr_cpu()->id;
+
+	for (p = 0; p < ARRAY_SIZE(_ready_q.q); p++) {
+		sys_dlist_t *list = &_ready_q.q[p];
+		sys_dnode_t *node;
+
+		for (node = list->tail; node != list; node = node->prev) {
+			struct k_thread *th = (struct k_thread *)node;
+
+			/* Skip threads that are already running elsewhere! */
+			if (th->base.active && th->base.cpu != mycpu) {
```

Comment:
what does active flag indicate? does it mean thread has run at least once? also when are we assigning the cpu? is it during thread create?

Reply:
Documented in the struct as "Non-zero when actively running on a CPU".  This indicates that another CPU is actively running this thread.

And we don't assign threads to CPUs.  The "S" in SMP means that every CPU can run every process.  They select the next process from the scheduler in this code when they are about to context switch (i.e. someone has called k_yield(), or we are returning from handling an interrupt of a preemptible thread, etc...).